README.md
# Akrophonobolos

Python package for handling Ancient Athenian monetary values in
talents, drachmas and obols, including input and output in Greek
acrophonic numerals (such as 𐅎, 𐅍, 𐅌, 𐅋, 𐅊, etc.)

The denominations of Ancient Athenian money were the _tálanton_
(τάλαντον, or talent), the _drakhmḗ_ (δραχμή, or drachma), and the
_obolós_ (ὀβολός, or obol). Six _oboloí_ made a _drakhmḗ_, and 6,000
_drakhmaí_ a _tálanton_ (which was 57 lbs. of silver).

Large sums of money are found in forms like “1 _tálanton_ 813
_drakhmaí_ 1½ _oboloí_.” Math with these figure can be very annoying,
so Akrophonobolos simpfifies this.

## Installation

    pip install akrophonobolos
    
## Usage

Akrophonobolos provides a class, `Khremata` (χρήματα, "money") and
function for manipulating these.

### Initializing

An instance of `Khremata` can be initialized in several different ways:

With a string that gives amounts with the abbreviations `t`, `d`, and
`b` or `o`:

    >>> import akrophonobolos as obol
    >>> obol.Khremata("1t 813d 1.5b")
    Khremata (1t 813d 1½b [= 40879.5 obols])

You can use upper or lowercase letters, and spaces do not matter:

    >>> obol.Khremata("1T 813D 1.5B")
    Khremata (1t 813d 1½b [= 40879.5 obols])

    >>> obol.Khremata("1t813d1.5b")
    Khremata (1t 813d 1½b [= 40879.5 obols])
    
You can use `o` for obols, but since this is too similar to a zero,
`b` is better:

    >>> obol.Khremata("1t 813d 1.5o")
    Khremata (1t 813d 1½b [= 40879.5 obols])

    >>> obol.Khremata("1T 813D 1.5O")
    Khremata (1t 813d 1½b [= 40879.5 obols])
	
Internally, the `Khremata` class stores the value as a (possibly
fractional) number of _oboloí_, and this number can be used directly
to initialize an instance:

    >>> obol.Khremata(40879.5)
    Khremata (1t 813d 1½b [= 40879.5 obols])
	
Finally you can use a string of [Unicode Greek acrophonic
numerals](https://en.wikipedia.org/wiki/Ancient_Greek_Numbers_(Unicode_block)):

    >>> obol.Khremata("Τ𐅅ΗΗΗΔ𐅂𐅂𐅂Ι𐅁")
    Khremata (1t 813d 1½b [= 40879.5 obols])
    
### Formatting

There are methods to format the value as an abbreviation, which is the
default string representation:

    >>> m = obol.Khremata("1t 813d 1.5b")
    >>> m.as_abbr()
    '1t 813d 1½b'
    >>> print(m)
    1t 813d 1½b
    
It can also be output as a phrase:

    >>> m.as_phrase()
    '1 talent, 813 drachmas, 1½ obols'
    
And as Greek numerals:

    >>> m.as_greek()
    'Τ𐅅ΗΗΗΔ𐅂𐅂𐅂Ι𐅁'
    
### Math

You can do basic math with instances of `Khremata`

    >>> obol.Khremata("1t") + obol.Khremata("3000d")
    Khremata (1t 3000d [= 54000.0 obols])
    
    >>> obol.Khremata("1t") - obol.Khremata("3000d")
    Khremata (3000d [= 18000.0 obols])
    
    >>> obol.Khremata("1t") * 2
    Khremata (2t [= 72000.0 obols])
    
    >>> obol.Khremata("1t") / 2
    Khremata (3000d [= 18000.0 obols])

Comparisons:

    >>> obol.Khremata("1t") == obol.Khremata("1t")
    True
    
    >>> obol.Khremata("1t") > obol.Khremata("3000d")
    True
    
    >>> obol.Khremata("1t") < obol.Khremata("3000d")
    False
    
Most of these operators work both between two instance of `Khremata`
and between a `Khremata` and anything that can be converted into a
`Khremata`:

    >>> obol.Khremata("1t") + "3000d"
    Khremata (1t 3000d [= 54000.0 obols])
   
    >>> obol.Khremata("1t") - "ΧΧΧ"
    Khremata (3000d [= 18000.0 obols])
    
    >>> obol.Khremata("1t") == 36000
    True
    
    >>> 18000.0 < obol.Khremata("1t")
    True
    
You cannot multiple two instances of `Khremata` since "talents
squared" does not have any meaning (this raises an
`UndefinedMonetaryOperation` error). If you divide a `Khremata` by a
`Khremata` though the units cancel out and the operation returns a
unitless `Fraction`:

    >>> obol.Khremata("1500d") / obol.Khremata("1t")
    Fraction(1, 4)
    
### Fractions, part 1

Above, we said that the `Khremata` class stores the value internally
as a (possibly fractional) number of _oboloí_. The more correct way to
state that is that internally, the `Khremata` stores the value
internally, in _oboloí_, as a Python
[Fraction](https://docs.python.org/3/library/fractions.html). You can
access this directly as the `b` property of the class. In many cases,
of course, this will be a unit fraction:

    >>> m = obol.Khremata("100t")
    >>> m.b
    Fraction(3600000, 1)
    
But monetary sums could be recorded down to the quarter-obol:

    >>> m = obol.Khremata("1t 1d 1.25b")
    >>> m.b
    Fraction(144029, 4) 
    
which is the `Fraction` form of 36,007.25 _oboloí_. Storing the value
as a `Fraction` avoids some issues with floating point math and better
approximates how Ancient Greeks did math, since they did not use
decimal numbers.


### Loans and Interest

Figures in _tálanta_, _drakhmaí_, and _oboloí_ are found in many
ancient Athenian inscriptions, and the most interesting of these
involve loans, such as the so-called "Logistai Inscription" ([IG I³
369](https://epigraphy.packhum.org/text/381)) which records loans
from the money held in the Parthenon and temples of other gods to the
Athenian state. Loans were made at simple interest, most commonly at
the rate of 1 _drakhmḗ_ per 5 _tálanta_ per day.

Akrophonobolos provides functions for working with loans. To start, you
can calculate a more useful version of the rate. Given an amount of
principal, a number of days, and an amount of interest to be returned,
you get back the amount of simple interest to be added for one day:

    >>> obol.interest_rate("5t", 1, "1d")
    Fraction(1, 30000)
    
That is, the simple interest rate is 1/30,000th of the principal per day.
    
For any loan, the amount of interest is simply the principal times the
rate times the term of the loan. If we borrowed 25 _tálanta_ for a
year at the common rate we would be expected to pay 1,825 _drakhmaí_ of
interest:

    >>> rate = obol.interest_rate("5t", 1, "1d")
    >>> obol.Khremata("25t") * rate * 365
    Khremata (1825d [= 10950.0 obols])
    
Of course Akrophonobolos has a function for this:

    >>> rate = obol.interest_rate("5t", 1, "1d")
    >>> obol.interest(obol.Khremata("25t"), 365, rate)
    Khremata (1825d [= 10950.0 obols])

And 1/30000th is the default rate, so you can leave it out if that's
the rate you're using:

    >>> obol.interest(obol.Khremata("25t"), 365)
    Khremata (1825d [= 10950.0 obols])
    
And instead of an instance of `Khremata` you can provide something that can be turned into a `Khremata`:

    >>> obol.interest("25t", 365)
    Khremata (1825d [= 10950.0 obols])
    
If you have the interest and the rate, you can use those to get the principal:

    >>> obol.principal("1825d", 365)
    Khremata (25t [= 900000.0 obols])
	
If you have the principal and the interest, you can get the loan
term, in days:

    >>> obol.loan_term("25t", "1825d")
    365
    
This last scenario is what we usually find in the inscriptions. For
instance, line 7 of the Logistai Inscription records one loan as

> 𐅊· τόκος τούτον ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ

or "50 _tálanta_. Interest on this 2 _tálanta_ 1,970 _drakhmaí_." We
can plug these values into `loan_term()` and see the the loan was for
1,397 days, just under 4 years:

    >>> obol.loan_term("𐅊", "ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ")
    1397
    
### Fractions, part 2: Rounding

Line 88 of the Logistai Inscription records another loan as 3,418
_drakhmaí_ 1 _obolós_, with interest of 1 _drakhmḗ_ 5½ _oboloí_:

> ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι, τόκος τούτο 𐅂ΙΙΙΙΙ𐅁

This loan, it turns out, was for just 17 days.

    >>> obol.loan_term("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι", "𐅂ΙΙΙΙΙ𐅁")
    17
    
Now, if we want to double-check this:

    >>> obol.interest("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι", 17)
    Khremata (1d 5¾b [= 11.75 obols])
    
We get an answer that is ¼ _obolós_ too high (11.75 instead of
11.5). We do not know how the ancient Greeks did this math, how they
rounded, or what kind of approximations they used. The smallest unit
they recorded was ¼ _obolós_, so in Akrohobolos the `interest()` and
`principal()` functions round up to this by default. You can get an
unrounded answer:

    >>> obol.interest("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι", 17, roundup=False)
    Khremata (1d 5b [= 11.621766666666666 obols])
    
We can see what the precise fraction is:
    
    >>> precise = obol.interest("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι", 17, roundup=False)
    >>> precise.b
    Fraction(1635618250918339, 140737488355328)
    
1,635,618,250,918,339/140,737,488,355,328ths is a quite a
fraction. Clearly the Greeks did some approximating. Maybe you can
play around with Akrophonobolos and figure out how they arrived at
11.5 obols for this amount.

`loan_term()` rounds to the nearest integer, but you can change this as well:

	>>> term = obol.loan_term("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι", "𐅂ΙΙΙΙΙ𐅁", roundoff=False)
    >>> term
    Fraction(345000, 20509)
    >>> float(term)
    16.82188307572285
    
## Command Line Scripts

Akrophonobolos provides two command line scripts: `obol` for
converting and simple math, and `logistes` for working with loans and
interest

### `obol`

If you give `obol` one or more amounts in either akrophonic numerals
or abbreviated with "t", "d" and "b" (or "o"), it will show the
equivalent forms

    $ obol 𐅉𐅉𐅈 348d "1d 5.5b" 14T1800D4O
    𐅉𐅉𐅈 = 25 talents
    348d = ΗΗΗΔΔΔΔ𐅃𐅂𐅂𐅂
    1d 5.5b = 𐅂ΙΙΙΙΙ𐅁
    14T1800D4O = 𐅉ΤΤΤΤΧ𐅅ΗΗΗΙΙΙΙ
    
You can also give `obol` numbers to add and subtract

    $ obol 1t + 1000d
    ΤΧ = 1t 1000d
    $ obol 1t - 1000d
    𐅆 = 5000d

### `logistes`

`logistes` will calculate principal, interest or loan terms based on
its inputs (`-p` for principal, `-i-` for interest, '`-d` for days of
loan):

    $ logistes -p 50t -d 1397
    𐅊 (50t) at 10 drachmas per day for 1397 days = ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ (2t 1970d) interest
    $ logistes -p 50t -i ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ
    𐅊 (50t) at 10 drachmas per day for 1397 days = ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ (2t 1970d) interest
    $ logistes -d 1397 -i ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ
    𐅊 (50t) at 10 drachmas per day for 1397 days = ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ (2t 1970d) interest
    
By default the rate is the common one, 5 _tálanta_ yield 1 _drakhmḗ_
in one day. You can change this with `--int-p`, `--int-i`, and
`--int-d`. To calculate the above at _2 drakhmaí_ per day per 5
_tálanta_:

    $ logistes -p 50t -d 1397 --int-p 5t --int-i 2d --int-d 1
    𐅊 (50t) at 20 drachmas per day for 1397 days = ΤΤΤΤΧΧΧ𐅅ΗΗΗΗΔΔΔΔ (4t 3940d) interest

## Contributing

Bug reports and pull requests are welcome on GitHub at
https://github.com/seanredmond/akrophonobolos



HISTORY.md
# Release History

## 0.0.2 (2021-11-22)

* Parse strings containing "½" and "¼" in the obols.

## 0.0.1 (2021-11-21)

Initial release


tests/test_class.py
import akrophonobolos as obol
from fractions import Fraction


def test_init():
    # Initialize with string

    money = obol.Khremata("1t")
    assert money.b == Fraction(36000, 1)

    money = obol.Khremata("1d")
    assert money.b == Fraction(6, 1)

    money = obol.Khremata("1b")
    assert money.b == Fraction(1, 1)

    money = obol.Khremata("1.5b")
    assert money.b == Fraction(3, 2)

    money = obol.Khremata("0.5b")
    assert money.b == Fraction(1, 2)

    money = obol.Khremata("0.25b")
    assert money.b == Fraction(1, 4)

    money = obol.Khremata("0.125b")
    assert money.b == Fraction(1, 8)
    
    money = obol.Khremata("1t 813d 1.5b")
    assert money.b == Fraction(81759, 2)

    # Initialized with Greek
    money = obol.Khremata("Τ𐅅ΗΗΗΔ𐅂𐅂𐅂Ι𐅁")
    assert money.b == Fraction(81759, 2)

    # Initialized with number
    money = obol.Khremata(40879.5)
    assert money.b == Fraction(81759, 2)

    money = obol.Khremata(40880)
    assert money.b == 40880
    
    money = obol.Khremata(163518.1)
    assert money.b == Fraction(5618439134432461, 34359738368)


def test_str():
    money = obol.Khremata("1t 813d 1.5b")
    assert f"{money}" == "1t 813d 1½b"


def test_repr():
    money = obol.Khremata("1t 813d 1.5b")
    print(f"MONEY: {money.__repr__()}")
    assert money.__repr__() == "Khremata (1t 813d 1½b [= 40879.5 obols])"


def test_as_abbr():
    money = obol.Khremata("1t 813d 1.5b")
    assert money.as_abbr() == "1t 813d 1½b"
    assert money.as_abbr(decimal=True) == "1t 813d 1.5b"


def test_as_greek():
    money = obol.Khremata("1t 813d 1.5b")
    assert money.as_greek() == "Τ𐅅ΗΗΗΔ𐅂𐅂𐅂Ι𐅁"


def test_as_phrase():
    money = obol.Khremata("1t 813d 1.5b")
    assert money.as_phrase() == "1 talent, 813 drachmas, 1½ obols"
    assert money.as_phrase(decimal=True) == "1 talent, 813 drachmas, 1.5 obols"


def test_int():
    money = obol.Khremata("1t 813d 1.5b")
    assert int(money) == 40879


def test_float():
    money = obol.Khremata("1t 813d 1.5b")
    assert float(money) == 40879.5


def test_eq():
    money = obol.Khremata("1t 813d 1.5b")
    assert money == 40879.5
    assert money == money
    assert money == obol.Khremata("1t 813d 1.5b")
    assert money == "1t 813d 1.5b"
    assert not money == 5
    assert not money == "5b"


def test_ne():
    money = obol.Khremata("1t 813d 1.5b")
    assert money != 163
    assert not money != money
    assert money != obol.Khremata("1t")


def test_add():
    money = obol.Khremata("5000d")
    assert money + money == "1t 4000d"
    assert isinstance(money + money, obol.Khremata)

    assert money + 1 == 30_001
    assert money + 1.5 == 30_001.5
    assert money + 1.1 == 30_001.1

    assert money + "5t" == "5t 5000d"
    assert money + "𐅈" == "5t 5000d"


def test_sub():
    money = obol.Khremata("5t 3000d")
    assert money - money == 0
    assert isinstance(money - "3000d", obol.Khremata)

    print(money - "3000d")
    assert money - "3000d" == "5t"
    assert money - "ΧΧΧ" == "5t"


def test_mul():
    money = obol.Khremata("5000d") * 2
    assert isinstance(money, obol.Khremata)
    assert money == "1t 4000d"


def test_div():
    money = obol.Khremata("1t 4000d") / 2
    assert isinstance(money, obol.Khremata)
    assert money == 30_000
    assert money.as_abbr() == "5000d"

    # a Khremata divided by a Khremata should return a Fraction
    assert money/money == 1
    assert isinstance(money/money, Fraction)

    assert obol.Khremata('2000d')/obol.Khremata('4000d') == Fraction(1, 2)


def test_interest():
    # Interest amounts from the beginning of IG I³ 369. Simple
    # interest of 1.2 obols per talent per day, or 1/30000th per day

    # IG I³ 369 l. 5-6
    principal = obol.Khremata("𐅉𐅉")
    days = 1424
    interest = (principal / 30000) * days
    assert interest.as_abbr() == "5696d"
    assert interest.as_greek() == "𐅆𐅅Η𐅄ΔΔΔΔ𐅃𐅂"

    # IG I³ 369 l. 7
    principal = obol.Khremata("𐅊")
    days = 1397
    interest = (principal / 30000) * days
    assert interest.as_abbr() == "2t 1970d"
    assert interest.as_greek() == "ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ"

    # IG I³ 369 l. 12
    principal = obol.Khremata("𐅋")
    days = 1197
    interest = (principal / 30000) * days
    assert interest.as_abbr() == "3t 5940d"
    assert interest.as_greek() == "ΤΤΤ𐅆𐅅ΗΗΗΗΔΔΔΔ"


tests/test_inscriptions.py
import akrophonobolos as obol
from fractions import Fraction
import pytest

def test_ig_I_3_369_7():
    # 𐅊· τόκος τούτον ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ
    assert obol.loan_term("𐅊", "ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ") == 1397
    assert obol.interest("𐅊", 1397) == obol.Khremata("ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ")
    assert obol.principal("ΤΤΧ𐅅ΗΗΗΗ𐅄ΔΔ", 1397) == obol.Khremata("𐅊")


def test_ig_I_3_369_7_8():
    # 𐅉[𐅉𐅈ΤΤΤ𐅆𐅅ΗΔΙΙΙ𐅁]· τόκος τούτον ΤΧ𐅅ΗΗΔ𐅃𐅂𐅂𐅂𐅂ΙΙ
    print(obol.Khremata("𐅉𐅉𐅈ΤΤΤ𐅆𐅅ΗΔΙΙΙ𐅁"))
    print(obol.Khremata("𐅉𐅉𐅈ΤΤΤ𐅆𐅅ΗΔΙΙΙ𐅁").b)
    print(obol.principal("ΤΧ𐅅ΗΗΔ𐅃𐅂𐅂𐅂𐅂ΙΙ", 1349))
    print(obol.principal("ΤΧ𐅅ΗΗΔ𐅃𐅂𐅂𐅂𐅂ΙΙ", 1349).b)
    assert obol.principal("ΤΧ𐅅ΗΗΔ𐅃𐅂𐅂𐅂𐅂ΙΙ", 1349) == "𐅉𐅉𐅈ΤΤΤ𐅆𐅅ΗΔΙΙΙ𐅁"


def test_ig_I_3_369_12():
    # 𐅋· τόκος τούτον ΤΤΤ𐅆𐅅ΗΗΗΗΔΔΔΔ
    assert obol.loan_term("𐅋", "ΤΤΤ𐅆𐅅ΗΗΗΗΔΔΔΔ") == 1197
    assert obol.interest("𐅋", 1197) == obol.Khremata("ΤΤΤ𐅆𐅅ΗΗΗΗΔΔΔΔ")
    assert obol.principal("ΤΤΤ𐅆𐅅ΗΗΗΗΔΔΔΔ", 1197) == obol.Khremata("𐅋")


def test_id_I_3_369_88():
    # ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι, τόκος τούτο 𐅂ΙΙΙΙΙ𐅁
    assert obol.loan_term("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι", "𐅂ΙΙΙΙΙ𐅁") == 17

    assert float(obol.loan_term("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι", "𐅂ΙΙΙΙΙ𐅁",
                                roundoff=False)) == 16.82188307572285

    assert obol.interest("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι", 17, roundup=False) == \
        11.621766666666666

    assert obol.principal("𐅂ΙΙΙΙΙ𐅁", 17, roundup=False) == 20294.11764705882
    

@pytest.mark.xfail(reason = "Different rounding")
def test_id_I_3_369_88_rounded_interest():
    # 1d 5¾b vs. 1d 5½b
    assert obol.interest("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι", 17) == obol.Khremata("𐅂ΙΙΙΙΙ𐅁")


@pytest.mark.xfail(reason = "Different rounding")
def test_id_I_3_369_88_rounded_principal():
    # 3382d 2¼b vs 3418d 1b
    assert obol.principal("𐅂ΙΙΙΙΙ𐅁", 17) == obol.Khremata("ΧΧΧΗΗΗΗΔ𐅃𐅂𐅂𐅂Ι")


akrophonobolos/akrophonobolos.py
from enum import IntFlag
from fractions import Fraction
import math
import re
from akrophonobolos.__version__ import __version__

class UnparseableMonetaryString(Exception):
    pass

class UndefinedMonetaryOperation(Exception):
    pass

AMT = re.compile(r"\A((\d+)T ?)?((\d+)D ?)?((\d+(\.\d+)?)(O|B))?\Z", re.I)
GREEK_AMT = re.compile(
    r"\A[\u0394\u0397\u0399\u03a4\u03a7\U00010140-\U0001014E]+\Z")


class Fmt(IntFlag):
    GREEK = 16
    ENGLISH = 8
    ABBR = 4
    DECIMAL = 2
    FRACTION = 1


# Amounts are stored internally obols and can be fractional
# 1 talent = 36,000 obols
# 1 drachma = 6 obols

NUMERALS = {"\U0001014E": Fraction(180_000_000, 1),  # 𐅎 5000 TALENTS
            "\U0001014D":  Fraction(36_000_000, 1),  # 𐅍 1000 TALENTS
            "\U0001014C":  Fraction(18_000_000, 1),  # 𐅌 500 TALENTS
            "\U0001014B":   Fraction(3_600_000, 1),  # 𐅋 100 TALENTS
            "\U0001014A":   Fraction(1_800_000, 1),  # 𐅊 FIFTY TALENTS
            "\U00010149":     Fraction(360_000, 1),  # 𐅉 TEN TALENTS
            "\U00010148":     Fraction(180_000, 1),  # 𐅈 FIVE TALENTS
            "\U000003a4":      Fraction(36_000, 1),  # Τ TAU (1 talent)
            "\U00010146":      Fraction(30_000, 1),  # 𐅆 FIVE THOUSAND
            "\U000003a7":       Fraction(6_000, 1),  # Χ KHI (1000 drachmas)
            "\U00010145":       Fraction(3_000, 1),  # 𐅅 FIVE HUNDRED
            "\U00000397":         Fraction(600, 1),  # Η ETA (100 drachmas)
            "\U00010144":         Fraction(300, 1),  # 𐅄 FIFTY
            "\U00000394":          Fraction(60, 1),  # Δ DELTA (10 drachmas)
            "\U00010143":          Fraction(30, 1),  # 𐅃 FIVE
            "\U00010142":           Fraction(6, 1),  # 𐅂 ONE DRACHMA
            "\U00000399":           Fraction(1, 1),  # Ι IOTA (1 obol)
            "\U00010141":           Fraction(1, 2),  # 𐅁 ONE HALF
            "\U00010140":           Fraction(1, 4)   # 𐅀 ONE QUARTER
            }

FMT_TDO = (NUMERALS["Τ"], NUMERALS["𐅂"])

# # Not used:
# # 10147 𐅇 GREEK ACROPHONIC ATTIC FIFTY THOUSAND


class Khremata():
    def __init__(self, amt, limit=None):
        self.b = self._parse_amt(amt, limit)

    def _parse_amt(self, amt, limit):
        if isinstance(amt, Fraction):
            if limit is None:
                return amt
            amt.limit_denominator(limit)
        
        if isinstance(amt, int):
            return amt * Fraction(4, 4)

        if isinstance(amt, float):
            if limit is None:
                return Fraction.from_float(amt) * Fraction(4, 4)
            return Fraction.from_float(amt).limit_denominator(limit)

        if isinstance(amt, Khremata):
            if limit is None:
                return amt.b
            return amt.b.limit_denominator(limit)

        if valid_greek_amount(amt):
            return parse_greek_amount(amt)

        if valid_amount_str(amt):
            return parse_amount(amt)

        raise UnparseableMonetaryString(
            f"Cannot parse {amt} as monetary amount")


    def as_abbr(self, decimal=False):
        if decimal:
            return format_amount(self.b, Fmt.ABBR | Fmt.DECIMAL)

        return format_amount(self.b, Fmt.ABBR)

    
    def as_greek(self):
        return format_amount(self.b.limit_denominator(4), Fmt.GREEK)


    def as_phrase(self, decimal=False):
        if decimal:
            return format_amount(self.b, Fmt.ENGLISH | Fmt.DECIMAL)

        return format_amount(self.b, Fmt.ENGLISH | Fmt.FRACTION)


    def __str__(self):
        return format_amount(self.b, Fmt.ABBR | Fmt.FRACTION)


    def __repr__(self):
        return (f"{self.__class__.__name__} ("
                f"{self.__str__()} [= {float(self.b)} obols])")


    def __int__(self):
        return int(self.b.limit_denominator(1))


    def __float__(self):
        return float(self.b)

    def __eq__(self, other):
        if isinstance(other, Khremata):
            return self.b == other.b

        # b (a Fraction) must be specifically converted to a float
        if isinstance(other, float):
            return float(self.b) == other
            
        return self.b == Khremata(other).b
    

    def __ne__(self, other):
        return self.b != other


    def __lt__(self, other):
        if isinstance(other, Khremata):
            return self.b < other.b

        return self.b < Khremata(other).b


    def ___le__(self, other):
        if isinstance(other, Khremata):
            return self.b <= other.b

        return self.b <= Khremata(other).b


    def __gt__(self, other):
        if isinstance(other, Khremata):
            return self.b > other.b

        return self.b > Khremata(other).b


    def __ge__(self, other):
        if isinstance(other, Khremata):
            return self.b >= other.b

        return self.b >= Khremata(other).b


    def __float__(self):
        return float(self.b)


    def __add__(self, other):
        if isinstance(other, Khremata):
            return Khremata(self.b + other.b)

        return Khremata(self.b + Khremata(other).b)


    def __sub__(self, other):
        if isinstance(other, Khremata):
            return Khremata(self.b - other.b)

        return Khremata(self.b - Khremata(other).b)


    def __mul__(self, other):
        if isinstance(other, Khremata):
            raise UndefinedMonetaryOperation("Cannot multiply two instances of"
                                             " Khremata")

        if isinstance(other, Fraction):
            return Khremata(self.b * other)

        return Khremata(self.b * float(other))


    def __truediv__(self, other):
        # The units cancel out when a Khremata id divided by a
        # Khremata, so return a Fraction
        if isinstance(other, Khremata):
            return Khremata(self.b / other.b).b

        # otherwise treat the divisor as a float and return an Khremata
        return Khremata(self.b / float(other))


    def __hash__(self):
        return hash(self.b)


def _qo(*amt):
    """ Convert tuple amount to quarter obols. """
    return amt[0] * Fraction(36_000, 1) + amt[1] * Fraction(6, 1) + Fraction(amt[2])


def rec_reduce(amt, denominations):
    """Recursively reduce obols to t/d/o. """
    if denominations:
        return (amt // denominations[0],) + \
            rec_reduce(amt % denominations[0], denominations[1:])

    return (amt,)


def valid_greek_amount(amt):
    return GREEK_AMT.search(amt) is not None


def valid_amount_str(amt):
    return AMT.search(amt) is not None


def parse_amount(amt):
    """ Parse an Athenian currency string into a tuple. """
    amt_match = AMT.match(amt)
    talents = 0 if amt_match[2] is None else int(amt_match[2])
    drachmas = 0 if amt_match[4] is None else int(amt_match[4])
    obols = 0 if amt_match[6] is None else float(amt_match[6])

    return _qo(talents, drachmas, obols)


def parse_greek_amount(amt):
    """ Parse Unicode Greek acrophonic numeral. """
    return sum([NUMERALS[c] for c in list(amt)])


def format_amount(amt, fmt_flags=Fmt.ABBR | Fmt.FRACTION):
    """ Format ¼-obols for readability """
    if fmt_flags & Fmt.GREEK:
        return "".join(_fmt_akrophonic(roundup_to_quarter_obol(amt)))

    if fmt_flags & Fmt.ENGLISH:
        return _fmt_tdo(rec_reduce(amt, FMT_TDO),
                        (("talent", "talents"), ("drachma", "drachmas"),
                         ("obol", "obols")),
                        _fmt_functions(fmt_flags),
                        " ", ", ")

    return _fmt_tdo(rec_reduce(amt, FMT_TDO),
                    (("t", "t"), ("d", "d"), ("b", "b")),
                    _fmt_functions(fmt_flags),
                    "", " ")


def interest_rate(p=Khremata("5t"), d=1, r=Khremata("1d")):
    """Calculate the simple interest rate that, given a principal amount
    p returns r in d days

    parameters:
        p  Khremata
        r  Khremata
        d  int

    Return value will be a fraction representing the amount of simple
    interest returned in one day, so that principal * rate * days will
    be the total amount of interest

    Default values return the common rate: 5 tálanta in on day returns
    1 drákhma.

    """

    if not isinstance(p, Khremata):
        return interest_rate(Khremata(p), d, r)

    if not isinstance(r, Khremata):
        return interest_rate(p, d, Khremata(r))
    
    return r/(p*d)


def interest(p, d, r=interest_rate(), roundup=True):
    """ 
    Calculate interest on principal p for d days at rate r

    Parameters
    p       Amount of principal. Can be an instance of Khremata or anything 
            that can be used to create an instance of Khremata
    d       Number of days over which to calculate interest
    r       Simple interest rate, should be an instance of Fraction (but can 
            be any number. Default value is the default returned by 
            interest_rate()
    roundup Boolean (default True). If True, result is rounded up to nearest 
            quarter obolós. If False, the exact amount is returned

    Return value will be an instance of Khremata.
    """

    if not isinstance(p, Khremata):
        return interest(Khremata(p), d, r, roundup)

    if roundup:
        return roundup_to_quarter_obol(p * r * d)

    return p * r * d


def loan_term(p, i, r=interest_rate(), roundoff=True):
    """ 
    Calculate loan term in days if principal was p and interest i at rate r

    Parameters
    p       Amount of principal. Can be an instance of Khremata or anything 
            that can be used to create an instance of Khremata
    i       Amount of interest. Can be an instance of Khremata or anything 
            that can be used to create an instance of Khremata
    r       Simple interest rate, should be an instance of Fraction (but can 
            be any number. Default value is the default returned by 
            interest_rate()

    """
    if not isinstance(p, Khremata):
        return loan_term(Khremata(p), i, r, roundoff)

    if not isinstance(i, Khremata):
        return loan_term(p, Khremata(i), r, roundoff)

    if roundoff:
        return round(i/(p*r))

    return i/(p*r)


def principal(i, d, r=interest_rate(), roundup=True):
    """ 
    Calculate the principal if loan returned i interest after d days at rate r

    Parameters
    i       Amount of interest. Can be an instance of Khremata or anything 
            that can be used to create an instance of Khremata
    d       Number of days from which to calculate principal
    r       Simple interest rate, should be an instance of Fraction (but can 
            be any number. Default value is the default returned by 
            interest_rate()
    roundup Boolean (default True). If True, result is rounded up to nearest 
            quarter obolós. If False, the exact amount is returned

    Return value will be an instance of Khremata.
    """
    if not isinstance(i, Khremata):
        return principal(Khremata(i), d, r, roundup)

    if roundup:
        return roundup_to_quarter_obol(i/(d * r))
    
    return i/(d * r)


def roundup_to_quarter_obol(o):
    """ 
    Roundup a value to the nearest quarter obol.

    Parameters
    o Value to be rounded (instance of Khremata, or any number)

    If passed an instance of Khremata, the return value will also be an 
    instance of Khremata. Otherwise the return value will be a float.
    """

    if isinstance(o, Khremata):
        return Khremata(roundup_to_quarter_obol(o.b))

    return math.ceil(o * 4)/4


def _fmt_akrophonic(amt):
    if not amt:
        return []

    num = [k for k, v in NUMERALS.items() if v <= amt][0]

    return [num] + _fmt_akrophonic(amt - NUMERALS[num])


def _fmt_fraction(amt):
    """ Format fractional obols as fractions. """
    if amt % 1 in (0.5, 0.25, 0.75):
        frac = {0.25: "¼", 0.5: "½", 0.75: "¾"}[amt % 1]
        whole = int(amt // 1) if amt // 1 else ""
        return f"{whole}{frac}"

    return int(amt)


def _fmt_decimal(amt):
    """ Format fractional obols as decimals. """
    if float(amt) % 1:
         return float(amt)

    return int(amt)


def _fmt_functions(fmt_flags):
    """ Return a tuple of functions to be used to format TDO. """
    if fmt_flags & Fmt.DECIMAL:
        return (int, int, _fmt_decimal)
    return (int, int, _fmt_fraction)


def _fmt_plural(amt, denominations):
    if amt <= 1:
        return denominations[0]

    return denominations[1]


def _fmt_tdo(tdo, denominations, fmt_funcs, delim1, delim2):
    return delim2.join(
        [f"{func(amt)}{delim1}{_fmt_plural(amt, d)}"
         for amt, d, func in zip(tdo, denominations, fmt_funcs) if amt])


def version():
    return __version__
